### PR TITLE
pageserver: finish vectored get early

### PR DESCRIPTION
**Pull request by VladLazar**
## Problem
If the previous step of the vectored left no further keyspace to investigate
(i.e. keyspace remains empty after removing keys completed in the previous step),
then we'd still grab the layers lock, potentially add an in-mem layer to the fringe and
at some further point read its index without reading any values from it.

## Summary of changes
If there's nothing left in the current keyspace, then skip the search and just select
the next item from the fringe as usual.

When running `test_pg_regress[release-pg16]` with the vectored read path for singular
gets this improved perf drastically:

skip_vs_no_skip_pg_regress

Legend:
blue - without this patch 
orange - with this patch

## Correctness
Since no keys remained from the previous range (i.e. we are on a leaf node) there's nothing
that search can find in deeper nodes.

## Checklist before requesting a review

- [X] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: pageserver/src/tenant/timeline.rs [hunk 1]
```diff
@@ -3127,55 +3127,57 @@ impl Timeline {
             unmapped_keyspace.remove_overlapping_with(&keys_done_last_step);
             completed_keyspace.merge(&keys_done_last_step);
 
-            let guard = timeline.layers.read().await;
-            let layers = guard.layer_map();
+            if unmapped_keyspace.start().is_some() {
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:3130]: 1. An `is_empty()` function would be more expressive. Let's add one?
2. There should be a brief comment explaining that this was shown to be important for performance.

File: pageserver/src/tenant/timeline.rs [hunk 2]
```diff
@@ -3127,55 +3127,57 @@ impl Timeline {
...
+                // The more interesting consideration is: "Why is the read algorithm still correct
+                // if the layer map changes while it is operating?". Doing a vectored read on a
+                // timeline boils down to pushing an imaginary lsn boundary downwards for each range
+                // covered by the read. The layer map tells us how to move the lsn downwards for a
+                // range at *a particular point in time*. It is fine for the answer to be different
+                // at two different time points.
+                drop(guard);
+            }
 
             if let Some((layer_to_read, keyspace_to_read, lsn_range)) = fringe.next_layer() {
```
Comment (problame) [inline pageserver/src/tenant/timeline.rs:3182]: Right above here, can we assert `unmapped_keyspace.is_empty()` , or did I forget again how this works?

**Review by problame — APPROVED** [review]

**Review by arpad-m — APPROVED** [review]
Good catch!

File: pageserver/src/tenant/timeline.rs (hunk 1 above)
Comment (arpad-m) [reply pageserver/src/tenant/timeline.rs:3130]: Yeah either a `KeySpace::is_empty` function or `unmapped_keyspace.ranges.is_empty()` would be nicer here.

File: pageserver/src/tenant/timeline.rs (hunk 2 above)
Comment (arpad-m) [reply pageserver/src/tenant/timeline.rs:3182]: The loop above only reads from `unmapped_keyspace.ranges`, it doesn't do a std::mem::take of the keyspace or such. So I don't think such an assertion would work.

Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs:3182]: Arpad is right. Asserting here doesn't really make sense. Let's map this to depth first search. Above line 3182 we expand the search space with the newly discovered vertices and below we pop a previously discovered vertex for processing.

File: pageserver/src/tenant/timeline.rs (hunk 1 above)
Comment (VladLazar) [reply pageserver/src/tenant/timeline.rs]: Good point: [e57853d](https://github.com/neondatabase/neon/pull/7490/commits/e57853decba2b268cdc137278f0ef9903bee323b)

**Pull request merged**